### ISSUE TITLE
vendoring: investigate tools to help with updating vendored deps

### ISSUE THREAD

**Issue by mpl** [perkeep/perkeep#889]
Title says it all. Maintaining and updating the (google) dependencies we vendor in has gotten painful enough that a tool to help with that would now be welcome.

I should investigate the existing options out there.

**Comment by cliveb** [comment]
Also interested in vendoring tools (working on a list). Will ask what Hyperledger ledger is using at the hackathon in NYC. https://medium.com/@ic/full-stack-development-and-go-26a64c70f387

**Comment by josephholsten** [comment]
I'm relatively happy with [govendor](https://github.com/kardianos/govendor)

**Comment by myitcv** [comment]
We follow a similar approach to [`vendo`](https://groups.google.com/d/msg/golang-nuts/GkE_BjSPocA/MneQU9lRCQAJ) using a very half-baked in-house tool (and plan to switch to `vendo` at some point). The general premise is:

* Create a directory called `_vendor` in the repo root, let's assume this is `/path/to/gopath/src/github.com/myitcv/project/_vendor`. This will contain our "vendored" dependencies (note the use of "vendored" vs. vendored; the latter is using the word vendor according to the definition [here](https://docs.google.com/document/d/1Bz5-UB7g2uPBdOx-rw5t9MxJwkfpx90cqG9AFL0JAYo/edit))
* When working on code, add `_vendor` to your `GOPATH` such that `GOPATH` is now `/path/to/gopath/src/github.com/myitcv/project/_vendor:/path/to/gopath`. We achieve this via [`smartcd`](https://github.com/cxreg/smartcd/) but I suspect there are other (better) ways
* A `go get` for any package will then be "vendored" into the `_vendor` directory
* Using `vendo` (or our half-baked in-house `simpkg`) the commit id of all the "vendored" code can then be captured in `_vendor/vendor.json` (or similar) at a package level (although strictly speaking we only need to capture this at the VCS level, the VCS repo containing any given package that is)
* As part of our CI, we loop through `_vendor/vendor.json` and ensure that all packages are at their respective commit ids; we then ensure our repo (git-based) is clean, i.e. that `_vendor/vendor.json` is consistent with the "vendored" code that has been committed. This only needs to happen during CI, and is predominantly in place to ensure that nobody has, by accident, committed a local change of some "vendored" code
* For programs that are part of our repo (all of these are, by convention, `cmd/*`), they vendor (in the Go 1.5 sense) their dependencies with a symlink to `_vendor/src`. Hence libraries and programs in our repo share the same "vendor"/vendor respectively

There are a couple of downsides to this approach:

* As far as I'm aware, there still exist constraints on using symlinks on Windows; put another way, on a Windows platform [supported by Go](https://golang.org/doc/install), you aren't guaranteed to have symlink support. This may be a deal-breaker for Camlistore developers... but I don't have any numbers to back this up (because the only relevant number is "how many Camlistore developers do we anticipate will use a Windows platform/setup that does not support symlinks?")
* If you do a `go get` on anything in the repo, all of the "vendored" dependencies are effectively downloaded twice; once in the clone of the repo in `_vendor`, and the second time because `go get` doesn't understand `_vendor` and so recursively gets the latest version of the dependencies of the repo. And also, if you're on a Windows box without symlink support, the vendor symlinks won't work. I [mooted the idea of "symlink" support in the `go` tool](https://github.com/golang/go/issues/15507) but, quite understandably, the proposal was declined; it's an area fraught with complexity cross-platform... but maybe this position will change

There are however a number of benefits; to list a few:

* It ensures consistent developer experience; whether working on a library or a program, everyone is always using the same code
* There is no versioning/semver nightmare to resolve/adjudicate
* Works for any VCS; the "vendored" code is a full copy of the entire repo containing a package
* ...

In case these thoughts help; feel free to simply discard if not 👍

**Comment by mpl** [comment]
it looks like it might be worth waiting for https://github.com/golang/dep to mature a bit too

**Comment by myitcv** [comment]
Updated my previous comment to correct a few points that were, at the time of writing this comment, out of date.

**Comment by myitcv** [comment]
@mpl can I make a suggestion whilst we await `dep` or another more permanent solution?

* Rename `/vendor` to be `/_vendor/src` - this will allow `_vendor` to be augmented to an existing `GOPATH`
* Maintain a `/.vendor.txt` file that contains three fields of data:

```bash
# format is: directory <TAB> commit_id <TAB> repo_url

/_vendor/src/github.com/gorilla/websocket	a91eba7f97777409bc2c443f5534d41dd20c5720	https://github.com/gorilla/websocket.git
/_vendor/src/github.com/myitcv/gopherjs		7f0a24fa1dfcd632237b79c2d55e21637bbbea7f	https://github.com/myitcv/gopherjs.git
....
```

* Maintain a `/_vendor/.gitignore` file to define what's included/excluded within vendored packages, e.g.:

```bash
# File /_vendor/.gitignore

/bin
/pkg
vendor/
_vendor/

# github.com/myitcv/gopherjs transforms
/github.com/myitcv/gopherjs/*
!/github.com/myitcv/gopherjs/react
!/github.com/myitcv/gopherjs/cmd
/github.com/myitcv/gopherjs/cmd/*
!/github.com/myitcv/gopherjs/cmd/reactGen

# ... etc for other packages

```

* The `/_vendor` directory can now, assuming remote repos are "available" and "behave well" (i.e. commits are retained), be entirely rebuilt at any point in time, by cloning each of the repos listed in `/.vendor.txt` into the listed directory, checking out to the listed commit, removing any `.git` and `.hg` etc directories created (keeps git commands in camlistore repo simpler), then running `git clean -x -f -d` within `/_vendor`
* Update CI-related checks to essentially rebuild `/_vendor` from `/.vendor.txt` and `/_vendor/.gitignore` and verify via `git status --porcelain` that there's no variation between the pinned reference and what's been committed

This will:

* Make it clear for any vendored dependency, which commit was the starting point, what transformations were applied (via `.gitignore`) to any given dependency at any point in time
* Allow all the vendored code to be easily referenced via a `GOPATH` (will allow for simpler development of camlistore, tools like `gocode` etc will then just work)
* Provide a well defined process by which new dependencies get added/repinned
* Provide an automated means by which we can detect if vendored dependencies were accidentally updated/modified by hand (the CI `git status --porcelain` check)
* Move away from Go 1.5 `vendor`-ing to `GOPATH` augmentation which is much cleaner/clearer

Thoughts?

**Comment by mpl** [comment]
This sounds like a lot of work to transition to something that, hopefully, we wouldn't keep for long since we would adopt dep soon. I'd rather switch to dep early on, and cope with whatever shortcomings it has with a couple of hacks.

Unless maybe your plan is already close to how dep works (I don't know, as I haven't played with dep enough so far), in which case it might be interesting to adopt it if it makes the later transition to dep smoother?

Would you be interested in drawing a comparison of your plan and of how dep works?

**Comment by myitcv** [comment]
> Would you be interested in drawing a comparison of your plan and of how dep works?

A very fair response... and a good challenge. 

Here's my _initial_ take on things as the stand. All comments as of https://github.com/golang/dep/commit/7f85e04e3446e25cac27cf5d07e4e56db3680ef0:

1. `dep`'s implementation is such that the `go` tool does not work for packages under its control (i.e. `go list`, `go test`, `go install`). This, to my mind, is a blocker for `dep`'s progress/adoption. I note this is no different from the drawbacks of using `vendor/` directories (which is unsurprising because `dep` uses `vendor/` directories under the hood) but I see this as a significant problem/unreasonable restriction for what is being touted as the official package management tool. (On a related note, this is exactly why I don't use `vendor/` directories and instead use augmented `GOPATH`'s) 
2. The workflow/tooling is focussed/oriented on dependencies as they exist in code, i.e. the transitive dependencies of a set of `.go` files. Whilst there is provision for manually specifying other dependencies, I'm yet to understand why there is such a focus. Because it gives rise to the rather odd workflow where you first have to write an import statement in code before then running `dep ensure ...` to fetch the code. Not a critical point per se and there is some acknowledgement that this might need to change, but I found this a strange initial decision and one that remains to date
3. There is a bug whereby `main` packages cannot be added as manual dependencies (but it's acknowledged this is a bug so will go away with time)
4. The current implementation appears to be rather consumed by the version resolution problem. To my mind this could/should be solved in `v2` of the tool. I'm trying to understand why this is considered critical for `v1`, seemingly at the expense of other rather more critical features/problems (all of the above points)
5. The format used for manifest and lock files looks set to change in the short term

In conclusion from my perspective:

* Point 1 blocking
* Point 2 very annoying 
* Point 3 blocking
* Point 5 a refactor I could simply avoid by waiting "some time"

So I'm not moving to use `dep` in any of my projects yet.

Your opinion on each of the above might differ from mine, particularly with respect to point 1 given that camlistore is currently using a `vendor/` directory. 

Which actually gives rise to the biggest difference: `dep` uses `vendor/` directories (whilst the is an implementation detail/decision, it has side effects in the form of point 1 hence why it's significant) whereas my approach is an augmented `GOPATH` approach. Quite how `dep` will solve point 1 remains to be seen... they might yet end up augmenting `GOPATH`s. Or they may choose not to deal with it at all... which to my mind would be a mistake.

This is a very initial, current take on things. I'm tracking discussion on all of the above so will try and post significant changes/updates back here as and when.

**Comment by mpl** [comment]
Hmm I thought pretty much everyone had agreed to adopt the <code>vendor/</code> directory convention. Or at least most of the popular (go) versioning tools out there, haven't they?

**Comment by tgulacsi** [comment]
Yes, they did. And the go tool itself accepted the `vendor/` folder, treats is specially, so anything  going against `vendor/` goes against the `go` tool, and has to go circles to achieve what vendor/ + go tool achieves.

**Comment by myitcv** [comment]
@mpl @tgulacsi ok to help me understand, what aspects/features of `vendor/` are we relying on in camlistore?

Comparing `vendor/` with a `GOPATH` approach, these are the two main differences I'm aware of:

* per above, the `go` tool does not work on `vendor/` "vendored" dependencies; I can't think we're relying on this :)
* `vendor/` has slightly different fall-through semantics where directories are empty; given that I _suspect_ the intent of using `vendor/` is to capture all of the dependencies of camlistore then we can't be relying on this either, i.e. it's an "error" for a dependency of camlistore not to be within `vendor/`

**Comment by lindner** [comment]
I've been able to get [govend](https://github.com/govend/govend) to mostly work.  The broken code.google.com repos cause problems so I'm just going to fix those manually.

**Comment by bradfitz** [comment]
Y'all are welcome to take this on early, but be ready to change it all again when the official Go support arrives, in whatever form that takes. I suspect it'll be different from the github.com/golang/dep tool, as Russ seems to have some plan.

**Comment by lindner** [comment]
The camlistore vendor tree is a bit of a mess, so any work here to track versions is going to help.

I've started the implementation of dep and it's a bit messy.  One issue is many subdirectories of whole repos were versioned, leading to inconsistencies.  There are also some local modifications for sqlite3 and a number of files that were omitted (which need to be added to a .gitignore)  Oh and the vendor/embed directory is totally against convention.

The Google Cloud deps are a horror; but the fork of grpc was dealt with somewhat well.

I'll have many more cleanups spawned from this effort, and maybe we'll get actual dep working.  But in the end any consistent vendoring solution beats what's there now.

There's also a lot of very very old deps there that need updating.

**Comment by lindner** [comment]
Okay, I have a working version with a dep-managed vendor directory.  Still need to work out the sqlite3 and fsnotify local mods, but it's pretty close.

The patch is over 1000 files though, so I'll introduce it in pieces as I rationalize our repos.

Here's the list of deps and pinned versions that I have enumerated:

```
PROJECT                                    CONSTRAINT  VERSION        REVISION  LATEST   PKGS USED
bazil.org/fuse                             *                          371fbbd            4
cloud.google.com/go                        *                          b70ccc7            8
github.com/FiloSottile/b2                  *                          b197f7a            1
github.com/bradfitz/latlong                *                          b745505            1
github.com/cznic/fileutil                  2d566d8     branch master  2d566d8   2d566d8  1
github.com/cznic/internal                  *                          4747030            3
github.com/cznic/kv                        *                          892ccf7            1
github.com/cznic/lldb                      bea8611     v1.1.0         bea8611   bea8611  1
github.com/cznic/mathutil                  09cde8d     branch master  09cde8d   09cde8d  1
github.com/cznic/sortutil                  4c73428     branch master  4c73428   4c73428  1
github.com/cznic/zappy                     2533cb5     branch master  2533cb5   2533cb5  1
github.com/edsrzf/mmap-go                  0bce6a6     branch master  0bce6a6   0bce6a6  1
github.com/fsnotify/fsnotify               *                          629574c            1
github.com/garyburd/go-oauth               *                          7d74905            1
github.com/go-sql-driver/mysql             *                          9a7aa36            1
github.com/golang/protobuf                 8ee7999     branch master  1e59b77            8
github.com/golang/snappy                   *                          553a641            1
github.com/googleapis/gax-go               *                          da06d19            1
github.com/gopherjs/gopherjs               *                          b40cd48            12
github.com/gopherjs/jquery                 *                          fbbfc4b            1
github.com/gopherjs/jsbuiltin              67703bf     branch master  67703bf   67703bf  1
github.com/gorilla/websocket               *                          2119675            1
github.com/hjfreyer/taglib-go              *                          0ef8bba            2
github.com/inconshreveable/mousetrap       *           v1.0           76626ae   76626ae  1
github.com/kardianos/osext                 *                          29ae4ff            1
github.com/kisielk/gotool                  *                          0de1eaf            1
github.com/lib/pq                          *                          9afcd9a            2
github.com/mailgun/mailgun-go              *                          17e8bd1            1
github.com/mattn/go-sqlite3                *                          919cf41            1
github.com/miekg/dns                       *                          3f1f7c8            1
github.com/neelance/astrewrite             *           branch master  9934826   9934826  1
github.com/neelance/sourcemap              *                          8c68805            1
github.com/nf/cr2                          *                          05d46fe            1
github.com/pkg/errors                      f15c970     v0.8.0         645ef00            1
github.com/plaid/plaid-go                  *                          02b6af6            1
github.com/russross/blackfriday            *                          151efb0            1
github.com/rwcarlsen/goexif                branch go1  branch go1     709fab3   709fab3  2
github.com/shurcooL/go                     *           branch master  004faa6   004faa6  1
github.com/shurcooL/httpfs                 *           branch master  809bece   809bece  1
github.com/shurcooL/sanitized_anchor_name  *                          11a20b7            1
github.com/spf13/cobra                     *                          c678ff0            1
github.com/spf13/pflag                     *                          7f60f83            1
github.com/syndtr/goleveldb                *                          4875955            12
github.com/syndtr/gosnappy                 *                          156a073            1
github.com/tgulacsi/picago                 *                          5f1a367            1
go4.org                                    *                          169ea6c            19
golang.org/x/crypto                        *                          ede567c            10
golang.org/x/image                         *                          b3d8467            4
golang.org/x/net                           *                          9dfe398            12
golang.org/x/oauth2                        *                          04e1573            5
golang.org/x/sync                          *                          fd80eb9            1
golang.org/x/sys                           *                          8fdfb00            1
golang.org/x/text                          *                          88f656f            27
golang.org/x/time                          *                          a4bde12            1
golang.org/x/tools                         *                          e531a2a            4
google.golang.org/api                      *                          48e49d1            15
google.golang.org/appengine                *           v1.0.0         150dc57   150dc57  12
google.golang.org/genproto                 *                          08f135d            10
google.golang.org/grpc                     *                          188a132            12
honnef.co/go/js/dom                        *                          24aa052            1
labix.org/v2/mgo                           *                          253                2
myitcv.io/gogenerate                       *                          bd69a94            1
myitcv.io/react                            *                          bca7c66            7
rsc.io/pdf                                 *                          1d34785            1
rsc.io/qr                                  *                          48b2ede            3
```

**Comment by mpl** [comment]
@lindner Not that I mind that you're updating all vendored deps to their latest revision (as long as it does not break anything of course), but I do not understand why that is necessary to start using the dep tool.
Why isn't it as "simple" as reorganizing a few paths, adding a few meta files, and some doc to explain what the steps are to use the dep tool in a workflow that suits us?
What's your plan here?

**Comment by lindner** [comment]
@mpl my plan is to slowly converge vendor in a safe manner via stepwise refinement.  I'm trying to keep things as close to the revisions in the repo, but sometimes I have to pull in new code to help satisfy the dependency tree.  (For example the official rev of picago had broken references, and x/net had some refs to camlistore that I want to get rid of)

So far the divergence has gone from 3500 modified files down to around 800.

There are some remaining challenges:

[ ] Local modifications to sqlite3
[ ] Google Cloud revs are diverged in the v0.3 -> v0.4 range
[ ] re-evaluate grpc http2 fork

And possibly others.

I can prepare a branch with all remaining changes so we can decide when to just finalize the remaining updates.

**Comment by mpl** [comment]
@lindner Could you show me what the tree would roughly look like once everything is in place, and more importantly, moving forward what would the workflow be for us to keep it updated in the future?

**Comment by lindner** [comment]
Sure!

You can see the dep patch (minus x/net) here that tries to match the versions what we have in vendor/ today:

https://camlistore-review.googlesource.com/c/camlistore/+/12670

[Note: We're going to need to move vendor/embed though, dep wants full control over the vendor directory.]

Once that's in we can modernize some of these three year old deps (with proper testing).

As for workflow the dep README has documenation:

https://github.com/golang/dep#updating-dependencies

Basically run dep and it will construct the transitive closure of all deps and find an acceptable vendor tree to match, adjust Gopkg.toml as needed.  Actual revisions are stored in the Gopkg.lock file.  Anything that touches vendor/ should be done by editing Gopkg.toml and rebuilding using *dep ensure*

**Comment by mpl** [comment]
@lindner ok, thanks, that helps!

**Comment by lindner** [comment]
Oh and we should also figure out the fsnotify issue for gopherjs.  I'll try
it without the custom patch and see if what affect it has on javascript
payload size.

**Comment by myitcv** [comment]
@lindner - `fsnotify` is only used by the GopherJS tool in its `serve` command so it doesn't have any impact on the resulting Javascript.

**Comment by mpl** [comment]
@lindner yes, I forgot to mention what Paul just said. we've been embedding a slightly hacked version of gopherjs everytime (basically breaking its server command), because Brad did not want to bother with fsnotify.

**Comment by myitcv** [comment]
Just a thought as a follow up to @bradfitz's [earlier comment](https://github.com/camlistore/camlistore/issues/889#issuecomment-345786193). Not to derail the current `dep` "pinning" (because formalising that is a good thing to my mind and will be transferrable) but more a repeat of a concern with `vendor`-based approaches in general. 

My biggest gripes with any `vendor`-based approach (`dep` included therefore) is that you lose use of the [`go` tool on your dependencies](https://github.com/golang/dep/issues/368). 

Consequently, as I've discussed with @mpl before, this makes it difficult to use tools that dependencies provide (e.g. `gopherjs` or `reactGen` that is a code generator to be used with `myitcv.io/react`). Amongst other things:

* Vendored tools have to be installed in a different way to the tool's docs: `go install perkeep.org/vendor/github.com/gopherjs/gopherjs`
* The installed tool then pollutes the `bin/` directory of the last entry in your `GOPATH` which means I need to isolate my Perkeep development entirely in a separate `GOPATH`

All of this seems to complicate the build and development process significantly to my mind, with no benefits.

Look forward to hearing more about Russ' thoughts on this in due course.

**Comment by mpl** [comment]
@lindner something is puzzling me: as I don't see any mention of dep in make.go, it is up to the developers to have dep installed, and in sync with all the other devs?

I mean, whenever I want to update something in vendor, I have to use dep, right? And so do all the other devs who want to do the same. So can we expect dep to stay stable enough that we don't care what version all the other devs are using, or can we expect some breaking changes that make some versions of dep incompatible with others?

If the latter, shouldn't we at least have some doc somewhere saying what version of dep the devs have to use? or better, have devcam, or something else, enforce it?

**Comment by lindner** [comment]
I added some docs to CONTRIBUTING.md  already.  I'm happy to expand on that.

Right now all versions of dep are working the same and result in the same generated vendor tree, so I haven't specified any specific versions.

Overall I view this as a temporary step to whatever dep management system go eventually adopts.

**Comment by mpl** [comment]
@lindner thanks. I had skimmed through it but missed that part. LGTM.

**Comment by mpl** [comment]
I've just tried the wipe, then rebuild with dep ensure; dep prune, looks like it's working nicely.
thanks for all that cleanup!

**Comment by mpl** [comment]
@lindner Hey. when i ran 'dep ensure && dep prune' while making this CL: https://camlistore-review.googlesource.com/c/camlistore/+/13486 , I ended up with 

	vendor/github.com/shurcooL/go/
	vendor/golang.org/x/tools/go/buildutil/
	vendor/golang.org/x/tools/refactor/

as untracked directories. It does not seem that they are needed since i can remove them and go run make.go seems to run. So what am I doing wrong here? Do I need to add them somewhere in the Gopkg.toml or not?

**Comment by lindner** [comment]
Thanks for spotting that.  The included packages are used by gopherjs tests, so that's why make.go still builds.

I'm inclined to keep them so that we have a deterministic state for the next developer that wants to change dependencies.

I've sent out https://camlistore-review.googlesource.com/#/c/camlistore/+/13546 to keep things synced.

Also note that running dep ensure could lead to changes in vendor if the constraints are not explicitly defined for packages.

**Comment by mpl** [comment]
Hmm, ok I hadn't seen your "deterministic state" argument here when I answered on the CL. As long as they are in .gitignore, I don't think a dev would notice they are generated, or be bothered by it, I dunno.

It would still be better if there was any way to specify in Gopkg.toml that we want some dirs to be force removed by dep prune. I assume it can't be done?

**Comment by lindner** [comment]
dep can ignore whole packages

https://github.com/golang/dep/blob/master/docs/Gopkg.toml.md#ignored

but I believe that this requires that the directory must have go code in it.  I can experiment with that and see if we can support pruning that way instead of .gitignore.

**Comment by mpl** [comment]
So are we now switching to vgo ASAP, or...? ;-)

**Comment by lindner** [comment]
vgo is built on top of dep, so we can certainly investigate how it can help
us.

**Comment by bradfitz** [comment]
vgo is certainly the answer long-term, but no need to rush into using its day 3 code. If dep is working now, let's keep using it a bit. vgo will support migrating from dep files later.
